collectors/google_pubsub/pubsub_collector.go

```diff
@@ -36,19 +60,143 @@ func (collector *PubSubCollector) HandleMessage(message *pubsub.Message) error {
+	ctx, cancel := context.WithCancel(context.Background())
+
+	// There's a built in deadline on the subscription that will orphan the message in the subscription.
+	subscription := collector.PubSubClient.Subscription(collector.SourceSubscription)
 	messages := make([]collectors.MessageWrapper, 0)
 
+	// Track number of messages received and timer to cancel this function when needed
+	found := make(chan int, 1)
+	received := 0
+	timer := time.NewTimer(5 * time.Second)
+	go func() {
+		for {
+			select {
+			case <- found:
+				received++
+				if received >= collector.PollingLimit {
+					log.Printf("Recieved %s messages\n", collector.PollingLimit)
+					cancel()
+				}
+			case <- timer.C:
+				log.Println("Timer fired")
+				cancel()
+			}
+		}
+	}()
+
+	// TODO - see if I need this
+	var mu sync.Mutex
+	err := subscription.Receive(ctx, func(ctx context.Context, msg *pubsub.Message) {
+		mu.Lock()
+		defer mu.Unlock()
+
+		log.Printf("Got message: %q\n", string(msg.Data))
+		// If value is not present in map we send an empty string in here, which will result in a 0 int.
+		retries, _ := strconv.Atoi(msg.Attributes["retries"])
+
+		messages = append(messages, collectors.MessageWrapper{
+			MessageBody: msg.Data,
+			Retries:     retries,
+			Retry:       false,
+			Fatal:       false,
+			Message:     msg,
+		})
+		msg.Ack()
+		found <- 1
+
+		// TODO - Create retry policy to use if processing fails
+		//  1. We could NACK it and then only ACK it when the buisness logic succeeds - this way we could use the
+		// 		delivery attempted attr on the msg with pubsub's inbound retry/dead letter queueing.
+		//		- Have to set up Dead letter queueing otherwise it will not count NACKs or failures
+		//  2. Roll my own retry counter.
+	})
+	if err != nil {
+		log.Printf("Could not recieve messages from - %s : %v", collector.SourceSubscription, err)
+	}
+	log.Printf("Get messages complete")
 	return messages, err
 }
 
 func (collector *PubSubCollector) PublishMessage(message *collectors.MessageWrapper, delaySeconds int64, errFlag bool) error {
-	var err error
-	return err
+	ctx := context.Background()
+	topic := collector.PubSubClient.Topic(collector.SourceTopic)
+	topicName := collector.SourceTopic
+	if errFlag {
+		topic = collector.PubSubClient.Topic(collector.ErrorTopic)
+		topicName = collector.ErrorTopic
+	}
+	// Clean up goroutines for batching and sending messages
+	defer topic.Stop()
+
+	message_attrs := make(map[string]string, 0)
+	message_retries := 0
+	if message.Retries >= 0 {
+		message_retries = message.Retries
+	}
+	message_attrs["retries"] = strconv.FormatInt(int64(message_retries), 10)
+
+	if delaySeconds > 0 {
+		time.Sleep(time.Duration(delaySeconds) * time.Second)
+	}
+
+	result := topic.Publish(ctx, &pubsub.Message{
+		Data: message.MessageBody,
+		Attributes: message_attrs,
+	})
+	// This will block until a result is returned from Google
+	id, err := result.Get(ctx)
+	if err != nil {
+		// TODO - need some way to retry publishes
```

Comment:
We can think later on to add a "publisher" to the "collector" to send messages and add that to the BaseCollector retry logic....  somehow....